ui/component/core/src/index.ts

```diff
@@ -1105,76 +962,118 @@ export class Manager implements EventProviderFactory {
     }
 
-    // When EventProvider (such as WebSocket) connection status changes.
-    // Always go OFFLINE if disconnected, only go back ONLINE when authentication is also connected.
-    protected _setEventDisconnected(disconnected: boolean, force = false) {
-        if(this._eventsDisconnected !== disconnected || force) {
-            console.debug(`EventProvider status changed: ${disconnected ? 'DISCONNECTED' : 'CONNECTED'}`);
-            if(disconnected) {
-                this._eventsDisconnected = true;
-                this._emitEvent(OREvent.OFFLINE);
-
-                // WebSocket might not connect because of token expiry, so we try updating it immediately.
-                // Since token expiry is unrelated to ONLINE/OFFLINE state, a successful token update would change _authDisconnected from FALSE to FALSE.
-                // So, if the token update is successful, we force update _authDisconnected.
-                if(this.isTokenExpired() && !this._authDisconnected) {
-                    this._tryUpdateAccessToken().then((offline) => {
-                        if(!offline) {
-                            this._setAuthDisconnected(false, true);
-                        }
-                    });
-                }
+    /**
+     * Checks keycloak is available and token is valid otherwise will redirect to login; also checks if event bus is
+     * online.
+     */
+    public async reconnect(reattemptDelayMillis: number = 3000) {
 
-            } else {
-                // When EventProvider is back online, check if authentication is disconnected and try to reauthenticate if necessary.
-                // Only emit the ONLINE event when both are connected.
-                this._eventsDisconnected = false;
-                if(this._authDisconnected) {
-                    this._runAuthReconnectTimer()
-                } else {
-                    this._emitEvent(OREvent.ONLINE);
+        if (!this._disconnected) {
+            return;
+        }
+
+        if (this._reconnectTimer) {
+            window.clearTimeout(this._reconnectTimer);
+            this._reconnectTimer = undefined;
+        }
+
+        const tryReconnect = async () => {
+            let keycloakOffline = !await this.isKeycloakReachable();
+
+            if (keycloakOffline) {
+                return false;
+            }
+
+            // Check if access token can be refreshed
+            try {
+                await this._updateKeycloakAccessToken();
+            } catch (e) {
+                // Try and use offline token if it is available
+                const offlineToken = await this._getNativeOfflineRefreshToken();
+                this._keycloak!.refreshToken = offlineToken;
+
+                try {
+                    await this._updateKeycloakAccessToken();
+                } catch (e) {
+                    this.login();
+                    return;
                 }
+                return true;
             }
```

Comment:
Yes, I like this 👍 

May be worth seperating `tryReconnect` to class function instead, but that's up to you.

Reply:
Happy to leave here as it is not used anywhere else and don't like creating spaghetti functions